commit.go

```diff
@@ -1,310 +1,199 @@
-			fmt.Fprintf(buf, "parent %s\n", hex.EncodeToString(cidToSha(p)))
-		}
-		fmt.Fprintf(buf, "author %s\n", c.Author.String())
-		fmt.Fprintf(buf, "committer %s\n", c.Committer.String())
-		if len(c.Encoding) > 0 {
-			fmt.Fprintf(buf, "encoding %s\n", c.Encoding)
-		}
-		for _, mtag := range c.MergeTag {
-			fmt.Fprintf(buf, "mergetag object %s\n", hex.EncodeToString(cidToSha(mtag.Object)))
-			fmt.Fprintf(buf, " type %s\n", mtag.Type)
-			fmt.Fprintf(buf, " tag %s\n", mtag.Tag)
-			fmt.Fprintf(buf, " tagger %s\n \n", mtag.Tagger.String())
-			fmt.Fprintf(buf, "%s", mtag.Text)
-		}
-		if c.Sig != nil {
-			fmt.Fprintln(buf, "gpgsig -----BEGIN PGP SIGNATURE-----")
-			fmt.Fprint(buf, c.Sig.Text)
-			fmt.Fprintln(buf, " -----END PGP SIGNATURE-----")
-		}
-		for _, line := range c.Other {
-			fmt.Fprintln(buf, line)
+func decodeCommitLine(c Commit, line []byte, rd *bufio.Reader) error {
+	switch {
+	case bytes.HasPrefix(line, []byte("tree ")):
+		sha, err := hex.DecodeString(string(line[5:]))
+		if err != nil {
+			return err
 		}
-		fmt.Fprintf(buf, "\n%s", c.Message)
-		c.rawData = buf.Bytes()
-	})
 
-	return c.rawData
-}
-
-func (c *Commit) Resolve(path []string) (interface{}, []string, error) {
-	if len(path) == 0 {
-		return nil, nil, fmt.Errorf("zero length path")
-	}
-
-	switch path[0] {
-	case "parents":
-		if len(path) == 1 {
-			return c.Parents, nil, nil
+		c.GitTree = _LinkTree{cidlink.Link{Cid: shaToCid(sha)}}
+	case bytes.HasPrefix(line, []byte("parent ")):
+		psha, err := hex.DecodeString(string(line[7:]))
+		if err != nil {
+			return err
 		}
 
-		i, err := strconv.Atoi(path[1])
+		c.Parents.x = append(c.Parents.x, _Link{cidlink.Link{Cid: shaToCid(psha)}})
+	case bytes.HasPrefix(line, []byte("author ")):
+		a, err := parsePersonInfo(line)
 		if err != nil {
-			return nil, nil, err
+			return err
 		}
 
-		if i < 0 || i >= len(c.Parents) {
-			return nil, nil, fmt.Errorf("index out of range")
+		c.Author = _PersonInfo__Maybe{m: schema.Maybe_Value, v: a}
+	case bytes.HasPrefix(line, []byte("committer ")):
+		com, err := parsePersonInfo(line)
+		if err != nil {
+			return err
 		}
 
-		return &node.Link{Cid: c.Parents[i]}, path[2:], nil
-	case "author":
-		if len(path) == 1 {
-			return c.Author, nil, nil
-		}
-		return c.Author.resolve(path[1:])
-	case "committer":
-		if len(path) == 1 {
-			return c.Committer, nil, nil
-		}
-		return c.Committer.resolve(path[1:])
-	case "signature":
-		return c.Sig.Text, path[1:], nil
-	case "message":
-		return c.Message, path[1:], nil
-	case "tree":
-		return &node.Link{Cid: c.GitTree}, path[1:], nil
-	case "mergetag":
-		if len(path) == 1 {
-			return c.MergeTag, nil, nil
+		c.Committer = _PersonInfo__Maybe{m: schema.Maybe_Value, v: com}
+	case bytes.HasPrefix(line, []byte("encoding ")):
+		c.Encoding = _String__Maybe{m: schema.Maybe_Value, v: &_String{string(line[9:])}}
+	case bytes.HasPrefix(line, []byte("mergetag object ")):
+		sha, err := hex.DecodeString(string(line)[16:])
```

Comment:
if you use a few global constants, then this gets easier, like `line[len(prefixMergetag):]`.